antelope_contracts/contracts/erc20/src/erc20.cpp

```diff
@@ -8,92 +8,269 @@
+        check(!!dest_acc, "destination address in bridge_message_v0 must be reserved address");
 
-    eosio::token::transfer_action transfer_act(token_account, {{get_self(), "active"_n}});
-    transfer_act.send(get_self(), *to, eosio::asset((uint64_t)(amount / minimum_natively_representable), token_symbol), std::string("Transfer from EVM"));
+        uint8_t amount_bytes[32]={};
+        memcpy(amount_bytes, (void *)&(msg.data[4 + kAddressLength]), 32);
+        
+        intx::uint256 value{}, zero{0_u256};
+        value = intx::be::load<intx::uint256>(amount_bytes);
+
+        intx::uint256 mult = intx::exp(10_u256, intx::uint256(itr->erc20_precision - itr->min_deposit.symbol.precision()));
+        check(value % mult == zero, "bridge amount can not have dust");
+        value /= mult;
+
+        uint64_t dest_amount = (uint64_t)value;
+        check(intx::uint256(dest_amount) == value && dest_amount < (1ull<<62)-1, "bridge amount value overflow");
+        check(dest_amount > 0, "bridge amount must be positive");
+
+        std::string memo;
+        int memo_len = (int)msg.data.size() - (4 + kAddressLength + 32);
+        if (memo_len > 0) {
+            memo.assign((const char *)&(msg.data[4 + kAddressLength + 32]), memo_len);
+        }
+
+        eosio::name dest_eos_acct(*dest_acc);
+        if (get_code_hash(dest_eos_acct) != checksum256()) {
+            egresslist_table_t(get_self(), get_self().value).get(dest_eos_acct.value, "native accounts containing contract code must be on allow list for egress bridging");
+        }
+
+        eosio::token::transfer_action transfer_act(itr->eos_contract_name, {{get_self(), "active"_n}});
+        transfer_act.send(get_self(), dest_eos_acct, eosio::asset(dest_amount, itr->min_deposit.symbol), memo);
+
+        token_table.modify(*itr, _self, [&](auto &v) {
+            v.balance -= dest_amount;
+        });
+    } else {
+        check(false, "unsupported bridge_message version");
+    }
 }
 
 void erc20::transfer(eosio::name from, eosio::name to, eosio::asset quantity,
                      std::string memo) {
-    eosio::check(get_first_receiver() == token_account && quantity.symbol == token_symbol,
-                 "received unexpected token");
 
     if (to != get_self() || from == get_self()) return;
 
-    if (memo.size() == 42 && memo[0] == '0' && memo[1] == 'x')
-        handle_evm_transfer(quantity, memo);
-    else
+    uint128_t v = get_first_receiver().value;
+    v <<= 64;
+    v |= quantity.symbol.code().raw();
+
+    token_table_t token_table(_self, _self.value);
+    auto index = token_table.get_index<"by.symbol"_n>();
+    auto itr = index.find(v);
+
+    eosio::check(itr != index.end() && itr->min_deposit.symbol == quantity.symbol, "received unregistered token");
+    eosio::check(quantity.amount >= itr->min_deposit.amount && quantity.amount > itr->deposit_fee.amount, "deposit amount too less");
```

Comment:
Only need the `quantity.amount > itr->deposit_fee.amount` check (though `deposit_fee` will be renamed to `ingress_fee`). Use error message `"deposit amount must be greater than ingress fee"`.